feat(ai): enable AI orchestrator discovery

rickstaa opened the pull request:
This commit incorporates the AIServiceRegistry contract address, superseding the conventional ServiceRegistry contract address. This strategic alteration streamlines the discovery process of AI Orchestrators within the AI Subnet, thereby bolstering network accessibility and interaction. While this approach serves as a swift workaround to enable the feature without extensive code modification, it's important to note that it may disrupt the existing transcoding discovery mechanism. We have to fix this if we want to merge the two networks in the future.

> [!WARNING]
> Do not merge into the main branch this pull request breaks the discovery of the regular transcoding network.

**What does this pull request do? Explain your changes. (required)**

<!-- A clear and concise description of what this pull request does. -->

This pull request implements https://linear.app/livepeer-ai-spe/issue/LIV-90/design-and-implement-solution-for-advertising-separate-service-uri-for so that AI subnet orchestrators can be found by the AI gateway.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->

-  `serviceRegistryAddr` was changed to a newly deployed [AIServiceRegistroyProxy](https://arbiscan.io/address/0x73f1755f34c2e769209e0f91a0c385722ed81b9c) contract. 

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

I checked if the service URIs of the orchestrators were to the values stored in the [AIServiceRegistroyProxy](https://arbiscan.io/address/0x73f1755f34c2e769209e0f91a0c385722ed81b9c) contract.  I have yet to test if an on-chain broadcaster can still send jobs.

**Does this pull request close any open issues?**

<!-- Fixes # -->

https://linear.app/livepeer-ai-spe/issue/LIV-90/design-and-implement-solution-for-advertising-separate-service-uri-for

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./CONTRIBUTING.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

rickstaa committed:
feat(ai): enable AI orchestrator discovery

This commit incorporates the AIServiceRegistry contract address,
superseding the conventional ServiceRegistry contract address. This
strategic alteration streamlines the discovery process of AI Orchestrators
within the AI Subnet, thereby bolstering network accessibility and interaction.
While this approach serves as a swift workaround to enable the feature without
extensive code modification, it's important to note that it may disrupt the
existing transcoding discovery mechanism. We have to fix this if we want to
merge the two networks in the future.

rickstaa commented:
Related to https://github.com/livepeer/protocol/pull/642. See my comments in https://github.com/livepeer/protocol/pull/642#issuecomment-2052699883.

rickstaa committed:
docs(ai): improve discovery documentation

This commit ensures that people are aware that they have to interact
with the `AIServiceRegistry` using their main Orch wallet.

yondonfu commented on eth/client.go:
```diff
@@ -211,7 +211,9 @@ func (c *client) setContracts(opts *bind.TransactOpts) error {
-	serviceRegistryAddr, err := c.GetContract(crypto.Keccak256Hash([]byte("ServiceRegistry")))
+	// TODO: Replace with call to controller once `AIServiceRegistry` is added to the controller.
+	// serviceRegistryAddr, err := c.GetContract(crypto.Keccak256Hash([]byte("ServiceRegistry")))
+	serviceRegistryAddr := ethcommon.HexToAddress("0x73f1755F34C2e769209E0F91a0c385722Ed81B9C")
```
I think hardcoding like this will also break usage of the node with local devnets or testnets since the node will try to use this hardcoded address on every single chain. One way to fix this is by only hardcoding the address if connected to Arbitrum One - something like:

```
chainID, err := c.backend.ChainID()
if err != nil { ... }

arbitrumOneChainID := big.NewInt(42161)
if chainID.Cmp(arbitrumOneChainID) == 0 {
    // Hardcode
} else {
    serviceRegistryAddr, err := c.GetContract(...)
}

...
```

See [this reference](https://docs.arbitrum.io/build-decentralized-apps/reference/node-providers) for Arbitrum chain IDs.

rickstaa replied:
Great point! I've made the change. I now have the non-upgradable `AIServiceRegistry`, which is hardcoded to the mainnet [controller](https://arbiscan.io/address/0xD8E8328501E9645d16Cf49539efC04f734606ee4). Additionally, I deployed a proxy contract that also links to the mainnet [controller](https://arbiscan.io/address/0xD8E8328501E9645d16Cf49539efC04f734606ee4) for our use. I believe both configurations are suitable; see my detailed comment in [this GitHub discussion](https://github.com/livepeer/protocol/pull/642#issuecomment-2062580312).

yondonfu commented on doc/ai-subnet.md:
````diff
@@ -369,6 +369,54 @@ The first and **recommended method** is to use the `ethOrchAddr` flag to set the
+    foundryup
+    ```
+
+2. Connect your _Mainnet Transcoding Network_ Orchestrator wallet to the Foundry cast using the following command:
````
I think an easier flow could be to use the `--keystore` and `--password` flags supported by [cast send](https://book.getfoundry.sh/reference/cli/cast/send?highlight=keystore#cast-send) since Os use keystore files to run go-livepeer already. 

Ex. to call `setServiceURI()`

```
cast send --keystore <PATH_TO_KEYSTORE_FILE> --password <PASSWORD> ... "setServiceURI(string)" ...
```

This would also avoid the need to directly deal with private keys since the user only needs to deal with keystore files.

rickstaa committed:
fix: fix 'AIServiceRegistry' devnet and testnet issue

This commit ensure that the hardcoded `AIServiceRegistry` contract
doesn't break the go-livepeer binary on local devnets or testnets.
